src/java/io/bazel/rulesscala/scalac/ScalaCInvoker.java

```diff
@@ -0,0 +1,366 @@
+        System.gc();
+      }
+    }
+  }
+
+  static private String[] extractSourceJars(CompileOptions opts, Path tmpParent) throws IOException {
+    List<File> sourceFiles = new ArrayList<File>();
+
+    for(String jarPath : opts.sourceJars) {
+      if (jarPath.length() > 0){
+        Path tmpPath = Files.createTempDirectory(tmpParent, "tmp");
+        sourceFiles.addAll(extractJar(jarPath, tmpPath.toString()));
+      }
+    }
+    String[] files = appendToString(opts.files, sourceFiles);
+    if(files.length == 0) {
+      throw new RuntimeException("Must have input files from either source jars or local files.");
+    }
+    return files;
+  }
+
+  private static List<File> extractJar(String jarPath,
+      String outputFolder) throws IOException, FileNotFoundException {
+
+    List<File> outputPaths = new ArrayList<File>();
+    java.util.jar.JarFile jar = new java.util.jar.JarFile(jarPath);
+    java.util.Enumeration e = jar.entries();
+    while (e.hasMoreElements()) {
+      java.util.jar.JarEntry file = (java.util.jar.JarEntry) e.nextElement();
+      File f = new File(outputFolder + java.io.File.separator + file.getName());
+
+      if (file.isDirectory()) { // if its a directory, create it
+        f.mkdirs();
+        continue;
+      }
+
+      File parent = f.getParentFile();
+      parent.mkdirs();
+      outputPaths.add(f);
+
+      java.io.InputStream is = jar.getInputStream(file); // get the input stream
+      java.io.FileOutputStream fos = new java.io.FileOutputStream(f);
+      while (is.available() > 0) {  // write contents of 'is' to 'fos'
+        fos.write(is.read());
+      }
+      fos.close();
+      is.close();
+    }
+    return outputPaths;
+  }
+
+  static <T> String[] appendToString(String[] init, List<T> rest) {
+    String[] tmp = new String[init.length + rest.size()];
+    System.arraycopy(init, 0, tmp, 0, init.length);
+    int baseIdx = init.length;
+    for(T t : rest) {
+      tmp[baseIdx] = t.toString();
+      baseIdx += 1;
+    }
+    return tmp;
+  }
+  public static String[] merge(String[]... arrays) {
+    int totalLength = 0;
+    for(String[] arr:arrays){
+      totalLength += arr.length;
+    }
+
+    String[] result = new String[totalLength];
+    int offset = 0;
+    for(String[] arr:arrays){
+      System.arraycopy(arr, 0, result, offset, arr.length);
+      offset += arr.length;
+    }
+    return result;
+  }
+
+  /**
+   * This is the reporter field for scalac, which we want to access
+   */
+  private static Field reporterField;
+  static {
+    try {
+      reporterField = Driver.class.getDeclaredField("reporter"); //NoSuchFieldException
+      reporterField.setAccessible(true);
+    }
+    catch (Exception ex) {
+      throw new RuntimeException("nope", ex);
+    }
+  }
+
+  private static void processRequest(List<String> args) throws Exception {
+    Path tmpPath = null;
+    try {
+      if (args.size() == 1 && args.get(0).startsWith("@")) {
```

Comment:
java doesn't pass args it consumes onto the runtime, so they never make it here if they are jvm_opts